File: src/github.rs

use eyre::{Context, Ok, Result};

use reqwest::{blocking::Client, header::{HeaderMap, HeaderValue}};
use serde_json::Value;
use crate::notification::Notification;

const GITHUB_URL: &str = "https://api.github.com/notifications"; 

pub fn get_notifications(client: &Client, token: &str) -> Result<Vec<Notification>> {
    let mut headers = HeaderMap::new();
    headers.insert("Authorization", HeaderValue::from_str(&format!("Bearer {}", token)).unwrap());
    headers.insert("Accept", HeaderValue::from_str("application/vnd.github+json").unwrap());
    headers.insert("X-GitHub-Api-Version", HeaderValue::from_str("2022-11-28").unwrap());
    headers.insert("User-Agent", HeaderValue::from_static("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"));
    let resp = client.get(GITHUB_URL).headers(headers).send().unwrap();
    let text = resp.text().unwrap();
    let data: Value = serde_json::from_str(&text).context("from_str")?;
    // let data: Value = resp.json().unwrap();
    let raw_notifications = data.as_array().unwrap();
    let mut notifications: Vec<Notification> = Vec::new();
    

    // Parse notifications
    for raw_notification in raw_notifications {
        let id = raw_notification["id"].as_str().unwrap().parse::<i128>().unwrap();
        let subject = &raw_notification["subject"];
        let title = subject["title"].as_str().unwrap_or("No Title");
        let url = subject["url"].as_str().map(|u| u.replace("api.github.com/repos/", "github.com/"));
        let kind = subject["type"].as_str().unwrap_or("Uknown Type").to_string(); 
        let respository = &raw_notification["repository"];
        let html_url = respository["html_url"].as_str().map(|s| s.to_string());
        let repo_name = respository["name"].as_str().map(|s| s.to_string());
        let updated_at = raw_notification["updated_at"].as_str().map(|s| s.to_string());
        let notification = Notification {id, title: title.into(), url, repo_name, kind, repo_url: html_url, updated_at};
        notifications.push(notification);
    }
    notifications.sort_by(|a, b| a.id.cmp(&b.id));
    Ok(notifications)
}

File: src/github.py
import requests
from typing import List, Dict, Optional
from notification import Notification

GITHUB_URL = "https://api.github.com/notifications"


def get_notifications(client: requests.Session, token: str) -> List[Notification]:
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"
    }
    
    response = client.get(GITHUB_URL, headers=headers)
    response.raise_for_status()  # This will raise an error for HTTP responses that are 4xx/5xx
    data = response.json()

    notifications = []
    
    for raw_notification in data:
        id = int(raw_notification["id"])
        subject = raw_notification["subject"]
        title = subject.get("title", "No Title")
        url = subject.get("url", "").replace("api.github.com/repos/", "github.com/") if subject.get("url") else None
        kind = subject.get("type", "Unknown Type")
        repository = raw_notification["repository"]
        repo_url = repository.get("html_url")
        repo_name = repository.get("name")
        updated_at = raw_notification.get("updated_at")
        
        notification = Notification(id, title, url, repo_name, kind, repo_url, updated_at)
        notifications.append(notification)

    notifications.sort(key=lambda x: x.id)
    return notifications
